### PR TITLE
Improve pagination for /markets /public APIs

### PR DESCRIPTION
**Pull request by mnaichuk**
- Add `api-pagination` gem for `account/deposits`, `account/withdraws`, `market/orders`, `market/trades`;
- Rewrite  unify params for `market/orders` and `market/trades`, `accounts/deposits` and `account/withdraws`;
- Sort by `desc` for `market/orders`;
- Rewrite `side` for `Trade` model;

### CODE REVIEW

**Comment by ysv** [comment]
use https://github.com/davidcelis/api-pagination and set pagination headers

**Review by ysv — CHANGES_REQUESTED** [review]
Use https://github.com/davidcelis/api-pagination instead

**Comment by ysv** [comment]
Related pr https://github.com/davidcelis/api-pagination/pull/119

File: app/api/v2/entities/trade.rb [hunk 1]
```diff
@@ -67,13 +67,13 @@ class Trade < Base
 
         expose(
           :side,
-          if: ->(trade, options) { options[:side] || trade.side },
+          if: ->(trade, options) { options[:side] || trade.side(options[:current_user]) },
```
Comment (ysv) [inline app/api/v2/entities/trade.rb]: `options[:side] || options[:current_user]`

File: app/api/v2/market/orders.rb [hunk 2]
```diff
@@ -16,17 +16,15 @@ class Orders < Grape::API
           optional :state, type: String, values: -> { Order.state.values }, desc: 'Filter order by state.'
           optional :limit, type: Integer, default: 100, range: 1..1000, desc: 'Limit the number of returned orders, default to 100.'
           optional :page,  type: Integer, default: 1, desc: 'Specify the page of paginated results.'
-          optional :order_by, type: String, values: %w(asc desc), default: 'asc', desc: 'If set, returned orders will be sorted in specific order, default to "asc".'
+          optional :order_by, type: String, values: %w(asc desc), default: 'desc', desc: 'If set, returned orders will be sorted in specific order, default to "asc".'
```
Comment (ysv) [inline app/api/v2/market/orders.rb]: `If set, returned orders will be sorted in specific order, default to "desc"``

File: app/api/v2/named_params.rb [hunk 3]
```diff
@@ -34,10 +34,9 @@ module NamedParams
       end
 
       params :trade_filters do
-        optional :limit,     type: Integer, range: 1..1000, default: 50, desc: 'Limit the number of returned trades. Default to 50.'
+        optional :limit,     type: Integer, range: 1..100, default: 100, desc: 'Limit the number of returned trades. Default to 50.'
```
Comment (ysv) [inline app/api/v2/named_params.rb]: `Limit the number of returned trades. Default to 50.` default is 100 now
Lets leave range: 1..1000 here

File: app/api/v2/account/withdraws.rb [hunk 4]
```diff
@@ -15,18 +15,17 @@ class Withdraws < Grape::API
         params do
           optional :currency, type: String,  values: -> { Currency.enabled.codes(bothcase: true) }, desc: -> { "Any supported currencies: #{Currency.enabled.codes(bothcase: true).join(',')}." }
           optional :page,     type: Integer, default: 1,   integer_gt_zero: true, desc: 'Page number (defaults to 1).'
-          optional :limit,    type: Integer, default: 100, range: 1..1000, desc: 'Number of withdraws per page (defaults to 100, maximum is 1000).'
+          optional :limit,    type: Integer, default: 100, range: 1..100, desc: 'Number of withdraws per page (defaults to 100, maximum is 1000).'
```
Comment (ysv) [inline app/api/v2/account/withdraws.rb]: update desc

File: config/initializers/kaminari_config.rb [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 Kaminari.configure do |config|
   config.default_per_page = 10
-  # config.max_per_page = nil
+  config.max_per_page = 100
```
Comment (ysv) [inline config/initializers/kaminari_config.rb]: put 1000

**Review by ysv — CHANGES_REQUESTED** [review]

File: app/api/v2/account/deposits.rb [hunk 6]
```diff
@@ -18,20 +18,19 @@ class Deposits < Grape::API
...
         get "/deposits" do
           currency = Currency.find(params[:currency]) if params[:currency].present?
 
-          current_user
+          deposits = current_user
             .deposits
             .order(id: :desc)
             .tap { |q| q.where!(currency: currency) if currency }
             .tap { |q| q.where!(aasm_state: params[:state]) if params[:state] }
-            .page(params[:page])
```
Comment (mod) [inline app/api/v2/account/deposits.rb:32]: Why page is not in SQL query ?

**Review by mod — CHANGES_REQUESTED** [review]

File: app/api/v2/account/deposits.rb (hunk 6 above)
Comment (mnaichuk) [reply app/api/v2/account/deposits.rb:32]: We will use `api-pagination` gem for return total amount in header and no need add `.page(params[:page])` in query.

Comment (ysv) [reply app/api/v2/account/deposits.rb:32]: @mod page is in SQL query 
Query is executed in `paginate` method

File: app/api/v2/public/markets.rb [hunk 7]
```diff
@@ -40,10 +40,12 @@ class OrderBook < Struct.new(:asks, :bids); end
...
-              limit: params[:limit], time_to: time_to,
-              from: params[:from], to: params[:to],
-              order: order_param)
+            trades = Trade.all
+              .order(order_param)
+              .tap { |q| q.where!(market: params[:market]) if params[:market] }
+              .tap { |q| q.where!(state: params[:state]) if params[:state] }
+              .page(params[:page])
+              .per(params[:limit])
             present trades, with: API::V2::Entities::Trade
```
Comment (ysv) [inline app/api/v2/public/markets.rb]: Why don't we use `present paginate(trades)` here ?

File: app/api/v2/account/withdraws.rb [hunk 8]
```diff
@@ -15,18 +15,17 @@ class Withdraws < Grape::API
...
-          current_user
+          withdraws = current_user
             .withdraws
             .order(id: :desc)
             .tap { |q| q.where!(currency: currency) if currency }
-            .page(params[:page])
-            .per(params[:limit])
-            .tap { |q| present q, with: API::V2::Entities::Withdraw }
+
+          present (paginate withdraws), with: API::V2::Entities::Withdraw
```
Comment (ysv) [inline app/api/v2/account/withdraws.rb]: Put this present into tap everywhere so we will have
Make this refactoring everywhere
```ruby
.tap { |q| q.where!(currency: currency) if currency }
.tap { |q| present paginate(q), with: API::V2::Entities::Withdraw }
```

File: config/initializers/api_pagination.rb [hunk 9]
```diff
@@ -0,0 +1,27 @@
+ApiPagination.configure do |config|
+  # If you have more than one gem included, you can choose a paginator.
+  config.paginator = :kaminari # or :will_paginate
+
+  # By default, this is set to 'Total'
+  config.total_header = 'X-Total'
```
Comment (ysv) [inline config/initializers/api_pagination.rb]: X - prefix in Headers is [deprecated](https://tools.ietf.org/html/rfc6648) so lets use default headers: Total, Per-Page ...
So you can remove this from initializer

File: config/initializers/api_pagination.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+  # By default, this is set to 'Per-Page'
+  config.per_page_header = 'X-Per-Page'
+
+  # Optional: set this to add a header with the current page number.
+  config.page_header = 'X-Page'
+
+  # Optional: set this to add other response format. Useful with tools that define :jsonapi format
+  config.response_formats = [:json, :xml, :jsonapi]
```
Comment (ysv) [inline config/initializers/api_pagination.rb]: Remove this line

File: config/initializers/api_pagination.rb [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  # Optional: what parameter should be used to set the page option
+  config.page_param = :page
+
+  # Optional: what parameter should be used to set the per page option
+  config.per_page_param = :limit
+
+  # Optional: Include the total and last_page link header
+  # By default, this is set to true
+  # Note: When using kaminari, this prevents the count call to the database
+  config.include_total = true
```
Comment (ysv) [inline config/initializers/api_pagination.rb]: if it's set to true by default don't add it one more time

**Review by ysv — CHANGES_REQUESTED** [review]

File: app/api/v2/public/markets.rb (hunk 7 above)
Comment (mnaichuk) [reply app/api/v2/public/markets.rb]: @ysv cause frontend team and R said that they don't need this feature

**Review by ysv — APPROVED** [review]

**Pull request merged**